QuadPlane: Allow separate FW landing approach radius and airspeed.

WickedShell opened the pull request:
This is the refinement of #9707. It allows quadplanes to have a separate radius when shooting a fixed wing approach to a `NAV_VTOL_LAND` waypoint, and includes  a separate land airspeed parameter. It has not been flown outside of the simulator yet. This does not provide any form of automatic radius calculation at the moment, as the existing `AP_Navigation::turn_distance()` function is unsuitable.

Other changes presented:
  - Fixed some whitespace that was bugging me
  - Pass the const reference to the mission command we already have for verifying the loiter commands rather then asking the mission library what it is.

Outside of real world flight testing things I'm considering on this (but haven't implemented):
  - Ensure that the selected radius is at least capable of giving us enough distance to do a into wind detransition + a minimum margin

pompecukor commented:
@wickedshell we are ready to test this live. Where is the radius defined? Thanks.

WickedShell commented:
@pompecukor the amazingly named `Q_FW_LND_APR_RAD` it's the radius in meters. I'm very open to better names for the parameter :)

pompecukor commented:
> 
> 
> @pompecukor the amazingly named `Q_FW_LND_APR_RAD` it's the radius in meters. I'm very open to better names for the parameter :)

OMG, hahaha.

I don't know how this work, I see "merging blocked" so I cannot test this in master right? What is the holdup?

WickedShell commented:
@pompecukor I'm not following. I just rebased this on master without any problems. (I pushed the rebased version up, so you could actually just build it off the branch if you wanted to)

WickedShell commented:
Added a fix to AP_Arming where the airspeed message was using the wrong print specifier, and made the indexing consistent with our other warnings.

WickedShell commented:
- Split the radius out of P1 as requested on the call
- Actually tested and fixed a lot of problems with CCW landings
- Added a general sanitization to L1 that the radius is positive. In general we should be managing this in the vehicle, but given the storage with negative values being CCW, it's actually fairly easy to accidentally pass a negative value to L1, which will go very badly from a control perspective.

pompecukor commented:
@WickedShell 
May I ask what the "different" landing air speed parameter is? Thank you. 
Also as a none coder, I really have no idea what "rebased" is. But I am happy to learn. :)

WickedShell commented:
Flight tested and worked as expected.

WickedShell commented:
@pompecukor Sorry I didn't answer you, thought I had... It uses `TECS_LAND_ARSPD`

pompecukor commented:
@wickedshell

Thank you dear. Hope to see it merged soon. :)

WickedShell commented:
@tridge flight tested, behaves as expected.

tridge commented on ArduPlane/commands_logic.cpp:
```diff
@@ -418,11 +418,18 @@ void Plane::do_land(const AP_Mission::Mission_Command& cmd)
+        loiter.direction = -1;
+    } else if (is_positive(quadplane.fw_land_approach_radius)) {
+        loiter.direction = 1;
+    }
```
what if radius is zero?

WickedShell replied:
Actually handled correctly by update_loiter. It's non intuitive so I've added a comment here.

tridge commented on ArduPlane/commands_logic.cpp:
```diff
@@ -971,21 +979,51 @@ bool Plane::verify_landing_vtol_approach(const AP_Mission::Mission_Command &cmd)
+
+                // validate that the vehicle is actually the expected distance away from the loiter point
+                // require an angle total of at least 2 centidegrees, due to special casing of 1 centidegree
+                if ((fabsf(get_distance(cmd.content.location, current_loc) - radius) > 7.5f) ||
```
maybe 10% of radius instead of 7.5m? Or 10% with a min of 7m?
I'm also a bit concerned that it could orbit forever if it can't quite make the radius

WickedShell replied:
Now accepts any larger radius as sufficient, the main point of the check is to ensure that we are a minimum distance from the target point.

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -377,6 +377,16 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
+    // @Range: 0 200
+    // @Increment: 10
+    // @User: Advanced
+    // @RebootRequired: True
```
why is reboot needed?

WickedShell replied:
Bad copy paste.

tridge commented on ArduPlane/quadplane.cpp:
```diff
@@ -377,6 +377,16 @@ const AP_Param::GroupInfo QuadPlane::var_info2[] = {
+    // @Description: This provides the radius used, when using a fixed wing landing approach. If set to 0 then the WP_LOITER_RAD will be selected.
+    // @Units: m
+    // @Range: 0 200
+    // @Increment: 10
```
why increment 10?

WickedShell replied:
Was arbitrary, reduced to 5, just trying to make cleaner sliders.
